File: packages/shared/src/engine/artifactNaming.ts
/**
 * FVL-04.026 — deterministic, cross-platform-safe naming for literature/
 * source documents and formulation export artifacts. Implements the
 * frozen contract in `docs/ARTIFACT_NAMING_SPEC.md`; the Python adapter
 * (`runtime/pipeline/artifact_naming.py`, literature naming only) passes
 * the SAME golden vectors (`artifactNaming.goldenVectors.json`) as this
 * file. No LLM/heuristic summarization anywhere here — every rule is a
 * fixed, deterministic string transform.
 *
 * Canonical database identity is never touched by anything in this
 * module: `Formulation.id`/`.code`, `FormulationVersion.id`/
 * `.versionNumber` are only ever READ here to build a derived display
 * string or filename — never renamed, never mutated.
 */

const WINDOWS_ILLEGAL = /[<>:"/\\|?*\x00-\x1F\x7F]/g;

const RESERVED_NAMES = new Set([
  "CON", "PRN", "AUX", "NUL",
  "COM1", "COM2", "COM3", "COM4", "COM5", "COM6", "COM7", "COM8", "COM9",
  "LPT1", "LPT2", "LPT3", "LPT4", "LPT5", "LPT6", "LPT7", "LPT8", "LPT9",
]);

/**
 * Sanitizes ONE filename component — never a display title, which stays
 * human Unicode text untouched. Deterministic: the same input always
 * produces the same output, on any run, on any platform.
 */
export function sanitizeFilenameComponent(raw: string, maxLength = 60): string {
  let s = (raw ?? "").normalize("NFC");
  s = s.replace(WINDOWS_ILLEGAL, "");
  s = s.replace(/\s+/g, " ").trim();
  s = s.replace(/ /g, "-");
  s = s.replace(/-{2,}/g, "-");
  s = s.replace(/^[-.]+|[-.]+$/g, "");
  if (s.length > maxLength) {
    s = s.slice(0, maxLength).replace(/[-.]+$/g, "");
  }
  if (RESERVED_NAMES.has(s.toUpperCase())) s = `${s}-file`;
  return s;
}

/**
 * Sanitizes a STABLE identifier component (a DOI or another external
 * source id) — DOI slashes map to `-` (kept visually recognizable, never
 * silently dropped) before the same illegal-character strip every
 * component gets. Never truncated by a caller — the stable id is what
 * makes two records with identical human-readable text collision-safe.
 */
export function sanitizeIdComponent(raw: string): string {
  let s = (raw ?? "").trim();
  s = s.replace(/\//g, "-");
  s = s.replace(WINDOWS_ILLEGAL, "");
  s = s.replace(/^[-.]+|[-.]+$/g, "");
  if (!s) return "UNKNOWN-ID";
  if (RESERVED_NAMES.has(s.toUpperCase())) s = `${s}-id`;
  return s;
}

// ------------------------------------------------------------ literature ---

export interface LiteratureNamingInput {
  firstAuthor?: string;
  year?: string | number;
  title: string;
  /** A DOI, or any other stable source identifier — never guessed, never
   *  derived from the title/author. */
  stableSourceId: string;
  /** Without the leading dot, e.g. "pdf". */
  extension: string;
}

/** Human-readable, NEVER sanitized — Unicode text for on-screen display,
 *  kept entirely separate from the physical filename. */
export function literatureDisplayTitle(input: Pick<LiteratureNamingInput, "firstAuthor" | "year" | "title">): string {
  const author = (input.firstAuthor ?? "").trim() || "Unknown Author";
  const year = String(input.year ?? "").trim() || "n.d.";
  const title = (input.title ?? "").trim() || "Untitled";
  return `${author} (${year}) — ${title}`;
}

const YEAR_RE = /^\d{4}$/;

export function literatureFilename(input: LiteratureNamingInput): string {
  const yearRaw = String(input.year ?? "").trim();
  const year = YEAR_RE.test(yearRaw) ? yearRaw : "UnknownYear";
  const authorRaw = (input.firstAuthor ?? "").trim();
  const author = authorRaw ? sanitizeFilenameComponent(authorRaw, 40) || "UnknownAuthor" : "UnknownAuthor";
  const shortTitle = sanitizeFilenameComponent(input.title ?? "", 60) || "Untitled";
  const id = sanitizeIdComponent(input.stableSourceId);
  const ext = (input.extension || "pdf").replace(/^\./, "").toLowerCase();
  return `LIT_${year}_${author}_${shortTitle}_${id}.${ext}`;
}

// ----------------------------------------------------------- formulation ---

/** Closed vocabulary — never arbitrary caller-supplied text. Derived from
 *  the real export types that exist in this codebase today (see
 *  `docs/ARTIFACT_NAMING_SPEC.md`'s own B1 audit table). */
export const FORMULATION_ARTIFACT_TYPES = [
  "Formula",
  "CostSnapshot",
  "PackagingBom",
  "ErpBom",
  "ErpRecipe",
  "Dossier",
  "EvidenceMatrix",
  "RegulatoryRules",
] as const;
export type FormulationArtifactType = (typeof FORMULATION_ARTIFACT_TYPES)[number];

export interface FormulationNamingInput {
  productFamily: string;
  formulaName: string;
  formulaCode: string;
  /** The formulation's own canonical version number — read only, never
   *  mutated or renamed. */
  version: number | string;
  artifactType: FormulationArtifactType;
  /** Without the leading dot, e.g. "xlsx". */
  extension: string;
}

/** Zero-padded 2-digit display (`V03`) for a real integer version number;
 *  a non-numeric/negative version falls back to a sanitized literal
 *  rather than fabricating a number. */
export function formulationVersionLabel(version: number | string): string {
  const n = typeof version === "number" ? version : Number.parseInt(String(version), 10);
  if (Number.isFinite(n) && n >= 0) return `V${String(n).padStart(2, "0")}`;
  return `V${sanitizeFilenameComponent(String(version), 10) || "00"}`;
}

/** Human-readable, NEVER sanitized. */
export function formulationDisplayTitle(input: Pick<FormulationNamingInput, "productFamily" | "formulaName" | "formulaCode" | "version">): string {
  return `${input.productFamily} — ${input.formulaName} — ${input.formulaCode} — ${formulationVersionLabel(input.version)}`;
}

export function formulationExportFilename(input: FormulationNamingInput): string {
  const family = sanitizeFilenameComponent(input.productFamily, 30) || "Unassigned";
  const name = sanitizeFilenameComponent(input.formulaName, 40) || "Formula";
  const code = sanitizeFilenameComponent(input.formulaCode, 30) || "NOCODE";
  const version = formulationVersionLabel(input.version);
  const ext = (input.extension || "xlsx").replace(/^\./, "").toLowerCase();
  return `FORM_${family}_${name}_${code}_${version}_${input.artifactType}.${ext}`;
}

File: runtime/pipeline/artifact_naming.py
"""FVL-04.026 -- deterministic, cross-platform-safe naming for literature
documents. Python mirror of `packages/shared/src/engine/artifactNaming.ts`,
implementing the SAME frozen contract (`docs/ARTIFACT_NAMING_SPEC.md`).
Both adapters pass the same golden vectors
(`packages/shared/src/engine/artifactNaming.goldenVectors.json`) -- proven
by `test_artifact_naming.py`.

Only literature naming is mirrored here: formulation export naming has no
Python caller anywhere in this codebase (every formulation export path is
TypeScript, `apps/desktop`), so it stays TypeScript-only per the "keep
language-specific adapters minimal" rule -- no unused Python surface.

No LLM/heuristic summarization anywhere in this module -- every rule is a
fixed, deterministic string transform, matching the TypeScript adapter
rule-for-rule.

Known cross-runtime limitation (disclosed, not hidden): truncation slices
by Unicode CODEPOINT here, matching TypeScript's UTF-16 CODE UNIT slicing
for every character in the Basic Multilingual Plane (everything this
literature pipeline's real sources produce). Neither adapter's golden
vectors exercise a character outside the BMP (e.g. an astral emoji), where
the two would diverge by at most one truncation boundary -- disclosed here
rather than solved, since no real input this pipeline handles needs it.
"""

from __future__ import annotations

import re
import unicodedata

_WINDOWS_ILLEGAL = re.compile(r'[<>:"/\\|?*\x00-\x1f\x7f]')

_RESERVED_NAMES = {
    "CON", "PRN", "AUX", "NUL",
    "COM1", "COM2", "COM3", "COM4", "COM5", "COM6", "COM7", "COM8", "COM9",
    "LPT1", "LPT2", "LPT3", "LPT4", "LPT5", "LPT6", "LPT7", "LPT8", "LPT9",
}

_YEAR_RE = re.compile(r"^\d{4}$")
_WS_RUN_RE = re.compile(r"\s+")
_HYPHEN_RUN_RE = re.compile(r"-{2,}")
_EDGE_SEP_RE = re.compile(r"^[-.]+|[-.]+$")
_TRAILING_SEP_RE = re.compile(r"[-.]+$")


def sanitize_filename_component(raw: str | None, max_length: int = 60) -> str:
    """Sanitizes ONE filename component -- never a display title, which
    stays human Unicode text untouched."""
    s = unicodedata.normalize("NFC", raw or "")
    s = _WINDOWS_ILLEGAL.sub("", s)
    s = _WS_RUN_RE.sub(" ", s).strip()
    s = s.replace(" ", "-")
    s = _HYPHEN_RUN_RE.sub("-", s)
    s = _EDGE_SEP_RE.sub("", s)
    if len(s) > max_length:
        s = _TRAILING_SEP_RE.sub("", s[:max_length])
    if s.upper() in _RESERVED_NAMES:
        s = f"{s}-file"
    return s


def sanitize_id_component(raw: str | None) -> str:
    """Sanitizes a STABLE identifier component (a DOI or another external
    source id) -- DOI slashes map to `-`, kept recognizable, never dropped."""
    s = (raw or "").strip()
    s = s.replace("/", "-")
    s = _WINDOWS_ILLEGAL.sub("", s)
    s = _EDGE_SEP_RE.sub("", s)
    if not s:
        return "UNKNOWN-ID"
    if s.upper() in _RESERVED_NAMES:
        s = f"{s}-id"
    return s


def literature_display_title(first_author: str | None, year: str | int | None, title: str | None) -> str:
    """Human-readable, NEVER sanitized -- kept separate from the physical filename."""
    author = (first_author or "").strip() or "Unknown Author"
    year_s = str(year if year is not None else "").strip() or "n.d."
    title_s = (title or "").strip() or "Untitled"
    return f"{author} ({year_s}) — {title_s}"


def literature_filename(
    first_author: str | None,
    year: str | int | None,
    title: str | None,
    stable_source_id: str,
    extension: str,
) -> str:
    year_raw = str(year if year is not None else "").strip()
    year_s = year_raw if _YEAR_RE.match(year_raw) else "UnknownYear"
    author_raw = (first_author or "").strip()
    author = (sanitize_filename_component(author_raw, 40) or "UnknownAuthor") if author_raw else "UnknownAuthor"
    short_title = sanitize_filename_component(title or "", 60) or "Untitled"
    source_id = sanitize_id_component(stable_source_id)
    ext = (extension or "pdf").lstrip(".").lower()
    return f"LIT_{year_s}_{author}_{short_title}_{source_id}.{ext}"
